07-algorithms/algorithms-divide-conquer.md

```diff
@@ -0,0 +1,388 @@
+1. If at any point, we end up with an empty range, we know the value was not in the array, and we can return a result indicating the value was not found, such as `None`. Other variations of binary search may return the index of where the value _should_ have been, but as a negative value to indicate that it was missing.
+1. Each recursion divides the array in half and performs the binary search on a smaller subproblem.
+
+![Performing binary search for the value 5 in the sorted list (1, 2, 3, 5, 6, 7, 8, 9). Looking for 5. Check the value at the midway index (s=0, m=4, e=8). 5 is less than 6. Adjust the ending index to rule out the right half. Check the value at the updated midway index (s=0, m=2, e=4). 5 is greater than 3. ADjust the starting index to rule out the left half. Check the value at the updated midway index (s=3, m=3, e=4). The value, 5, was found. Return the position, 3.](images/algorithmic-strategies_divide-and-conquer_binary-search.png)  
+_Fig. Looking for the position of the value 5 in a sorted array using binary search._
+
+<br />
+
+<details>
+
+<summary>We can view a recursive implementation of binary search by expanding this section.</summary>
+
+```python
+def recursive_binary_search(array, to_find, low=0, high=None):
+    if high is None:
+        high = len(array)
+
+    if high <= low:
+        return None
+
+    mid = (high + low) // 2
+
+    if array[mid] == to_find:
+        return mid
+    elif array[mid] > to_find:
+        high = mid
+    else:
+        low = mid + 1
+
+    return recursive_binary_search(array, to_find, low, high)
+```
+
+</details>
+
+## Check for Understanding
+
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+<!-- Replace everything in square brackets [] and remove brackets  -->
+
+### !challenge
+
+* type: multiple-choice
+* id: 645bad5d-af5f-42a9-a7e4-c70412972ab1
+* title: Time Complexity of Quicksort
+
+##### !question
+
+What is the *Worst-Case* time complexity for Quicksort?
+
+##### !end-question
+
+##### !options
+
+a| O(NlogN)
+b| O(N)
+c| O(N<sup>2</sup>)
+
+##### !end-options
+
+##### !answer
+
+c|
+
+##### !end-answer
+
+
+##### !explanation
+
+The worst-case time complexity for Quicksort is O(N<sup>2</sup>) because picking the worst pivot (e.g. the smallest or biggest remaining element) for each step will result in one array of size n-1 as opposed to two arrays that are approximately n/2 in length.
+
+##### !end-explanation 
+
+### !end-challenge
+
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+<!-- Replace everything in square brackets [] and remove brackets  -->
+
+### !challenge
+
+* type: checkbox
+* id: 25522151-56d9-4495-9691-da53700f896e
+* title: Divide and Conquer Algorithms
+
+##### !question
+
+Select all of the following which would be considered Divide/Decrease and Conquer algorithms.
+
+##### !end-question
+
+##### !options
+
+a| Mergesort
+b| Dijkstra's Algorithm
+c| Kadane's Algorithm
```

Comment:
Maybe change this to Insertion or Selection sort? Kadane's algorithm will be introduced after.

Reply:
Good callout! Updated with Insertion sort and added a link to more info about the technique, just in case it's their first time encountering the algorithm. Thank you!